---
title: "Differential Expression Analysis"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Differential Expression Analysis}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

```{r setup}
library(ngstan)
library(posterior)
```

# Introduction 

In this vignette, we'll replicate a portion of the 
[example differential expression analysis](https://github.com/hbctraining/Training-modules/blob/master/archive/DESeq2/lessons/03_DEG_deseq2.md)
from the Harvard Chan Bioinformatics Core, using data from [Kenny, et al.](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE50499). The package
comes with a filtered version of this data set 
(genes that average less than 2 counts per million across samples are filtered out), 
called `mov10`. 

```{r}
head(mov10)
```

Below, we'll set up a differential expression analysis that compares
the Mov10 Overexpression samples (`high_1`, etc.) to the control samples 
(`control_1`, etc.). 

# Preparing the model inputs

## Specifying the experimental design

We represent the experimental design as a model matrix. Since we're focusing
on a single treatment condition (Mov10 Overexpression), this is just a 
two-column matrix.

```{r}
# the first three samples are in the treatment condition
mov10_overexpression <- c(1, 1, 1, 0, 0, 0)
X_g <- model.matrix(~mov10_overexpression)
print(X_g)
```


## Identifying sharp nulls

If we want to test sharp null hypotheses (i.e. there is literally 
no difference in 
expression between treatment and control for gene $g$), we need to identify the
model parameters that are associated with the sharp null hypotheses so that 
the model uses a mixture prior for those parameters. 

We do this with an indicator vector: 

```{r}
mixture_probabilities <- c(1, 0.2)
```

The order corresponds to the order of columns in the model matrix. We are not 
interested in testing whether or not the log-scale intercept is exactly 0, so
the first element of `mixture_probabilities` is 0 - the model will not use a mixture
prior for the intercept. Since we would like to test
$$
H_0^g: \text{there is no difference in expression level between treatment
and control for gene } g,
$$
we use a mixture prior for the parameter associated with the second column of 
the model matrix - the parameter that represents the difference in (log) 
expression level between treatment and control. 

## Arguments list

```{r}
args <- list(
  y = as.matrix(mov10[, 2:ncol(mov10)]), # don't want geneid
  G = nrow(mov10), # number of genes / tags
  X_g = X_g,
  Z_g = matrix(data = numeric(), ncol = 0),
  Z_g = diag(ncol = 6, nrow = 6),
  mixture_probabilities = mixture_probabilities
)
```

# Running the model

As an aside, this model is huge and will take a long time to sample. 
If you have access to a computing cluster, I would recommend setting up
a model fitting and analysis pipeline to run on your cluster. See the 
vignette `"Computational costs"` for more information about how to minimize
model sampling time and how to use the alternatives to MCMC sampling. 

```{r eval=FALSE}
debugonce(run_hpl_glmm_mix_model)
fit <- do.call(run_hpl_glmm_mix_model, args)
```


# Analyzing the model output

```{r eval=FALSE}
draws <- fit$draws()
```
